File: DP/rod_cutting_max_profit.py
# import sys
# sys.setrecursionlimit(10**6)

def profit_t_d(n,cost,dp): # Time Complexity = O(n2)
	if n==0:
		return 0

	if dp[n]!=0:
		return dp[n]

	best = 0
	for i in range(1,n+1):
		curr_profit = cost[i] + profit_t_d(n-i,cost,dp)
		best = max(curr_profit,best)
	dp[n] = best
	return dp[n]

def profit_b_u(n,cost): # Time Complexity = O(n2)
	dp = [0]*(n+1)
	for length in range(1,n+1):
		best = 0
		for cut in range(1,length+1):
			best = max(best,cost[cut]+dp[length-cut])
		dp[length] = best
	return dp[n]


def main():
	n = 4 	
	dp = [0]*(n+1)
	cost = [0,2,3,2,5]
	print(profit_t_d(n,cost,dp))
	print(profit_b_u(n,cost))
main()

File: DP/rod_cutting_max_profit.cpp
#include<bits/stdc++.h>
using namespace std;

#define ll long long

int dp[100];


int profit_t_d(int n,int cost[])
{
	if(n==0)
		return 0;

  if(dp[n]!=0)
    return dp[n];

	int best = 0;
	for(int i=1;i<=n;i++)
	{
		int ans = cost[i] + profit_t_d(n-i,cost);
		best = max(best,ans);
	}
 	dp[n] = best;
	return dp[n];
}

int profit_b_u(int n,int cost[])
{
	int dp[100] = {0};
	
	for(int len=1;len<=n;++len)
	{
		int best = 0;
		for(int k=1;k<=len;++k)
		{
			best = max(best,cost[k] + dp[len-k]);
		}
		dp[len] = best;
	}	
	return dp[n];
}

int main()
{
	int len;
	cin>>len;
	int cost[len];

	for(int i=1;i<=len;i++)
		cin>>cost[i];

	cout<<profit_t_d(len,cost)<<endl;
	cout<<profit_b_u(len,cost)
}
